pkg/scheduler/scheduler.go

```diff
@@ -456,21 +418,93 @@ func (s *Scheduler) doScheduleClusterBinding(name string) (err error) {
+	defer klog.V(4).InfoS("End scheduling resource binding with ClusterAffinity", "resourceBinding", klog.KObj(rb))
 
-	scheduleResult, err := s.Algorithm.Schedule(context.TODO(), &resourceBinding.Spec, &core.ScheduleAlgorithmOption{EnableEmptyWorkloadPropagation: s.enableEmptyWorkloadPropagation})
+	scheduleResult, err := s.Algorithm.Schedule(context.TODO(), &rb.Spec, &core.ScheduleAlgorithmOption{EnableEmptyWorkloadPropagation: s.enableEmptyWorkloadPropagation})
 	var noClusterFit *framework.FitError
 	// in case of no cluster fit, can not return but continue to patch(cleanup) the result.
 	if err != nil && !errors.As(err, &noClusterFit) {
-		klog.Errorf("Failed scheduling ResourceBinding %s/%s: %v", resourceBinding.Namespace, resourceBinding.Name, err)
+		s.recordScheduleResultEventForResourceBinding(rb, err)
+		klog.Errorf("Failed scheduling ResourceBinding(%s/%s): %v", rb.Namespace, rb.Name, err)
 		return err
 	}
 
-	klog.V(4).Infof("ResourceBinding %s/%s scheduled to clusters %v", resourceBinding.Namespace, resourceBinding.Name, scheduleResult.SuggestedClusters)
-	scheduleErr := s.patchScheduleResultForResourceBinding(resourceBinding, placementStr, scheduleResult.SuggestedClusters)
-	return utilerrors.NewAggregate([]error{err, scheduleErr})
+	klog.V(4).Infof("ResourceBinding(%s/%s) scheduled to clusters %v", rb.Namespace, rb.Name, scheduleResult.SuggestedClusters)
+	patchErr := s.patchScheduleResultForResourceBinding(rb, placementStr, scheduleResult.SuggestedClusters)
+	s.recordScheduleResultEventForResourceBinding(rb, utilerrors.NewAggregate([]error{err, patchErr}))
+
+	// only care about the patch result,
+	// for FitError already recorded by event
+	return patchErr
+}
+
+func (s *Scheduler) scheduleResourceBindingWithClusterAffinities(rb *workv1alpha2.ResourceBinding, placementStr string) error {
+	klog.V(4).InfoS("Begin scheduling resource binding with ClusterAffinities", "resourceBinding", klog.KObj(rb))
+	defer klog.V(4).InfoS("End scheduling resource binding with ClusterAffinities", "resourceBinding", klog.KObj(rb))
+
+	beginAffinityIndex := getAffinityIndex(rb.Spec.Placement.ClusterAffinities, rb.Status.SchedulerObservedAffinityName)
+	affinityIndex := beginAffinityIndex
+
+	var scheduleResult core.ScheduleResult
+	var err, firstErr error
+	for affinityIndex < len(rb.Spec.Placement.ClusterAffinities) {
+		affinityName := rb.Spec.Placement.ClusterAffinities[affinityIndex].AffinityName
+		scheduleOption := &core.ScheduleAlgorithmOption{
+			SchedulerObservedAffinityName:  affinityName,
+			EnableEmptyWorkloadPropagation: s.enableEmptyWorkloadPropagation,
+		}
+		klog.V(4).Infof("Schedule ResourceBinding(%s/%s) with affinityName(%s)", rb.Namespace, rb.Name, affinityName)
+		scheduleResult, err = s.Algorithm.Schedule(context.TODO(), &rb.Spec, scheduleOption)
+		if err == nil {
+			break
+		}
+
+		// obtain to err of the first scheduling
+		if firstErr == nil {
```

Comment:
This will be the first failed scheduling.
If the first failed scheduling is non-noClusterFit, and the second failed scheduling is noClusterFit. The logic looks like it will just return an error, is this expected?

Reply:
If the first failed scheduling is `noClusterFit` and the second failed scheduling is `non-noClusterFit`.

This logic would seem to ignore the second error.

Reply:
Yes. IMOP, if no affinity is scheduled successfully, the affinityName in the binding state does not change. In this case, the error pointed to should be the error in the first affinity.